### PR TITLE
Add an Async Storage plugin

### PR DESCRIPTION
**Pull request by jthoward64**
Adds a plugin for `@react-native-async-storage/async-storage`

Features:

- [x] Show the full contents of AsyncStorage in a table
- [x] Edit and delete entries
- [x] Add new entries
- [x] Tested on iOS
- [ ] Tested on Android
- [x] Tested on Web
- [x] JSON formatted viewing/editing
- [x] High quality UI
- [ ] ~Seamless refreshing~

Devtools GUI:
image

Example app:
image

### CODE REVIEW

**Comment by jthoward64** [comment]
I think I'm either going to replace the entire UI with something more like the tinybase UI, or make it so the table conditionally renders formatted JSON instead of a text box. The UI isn't terrible, but could use some work (namely better sizing), and like the point above, a table might not be the best option. Refreshing is complicated, AsyncStorage doesn't have a way to listen for changes, so automatic updates are a non-option. Really the only path forward is polling for new changes and improving the in-progress editing support.

**Comment by Kudo** [comment]
hey! that's awesome for doing such a great work even we didn't document much how to create a plugin yet.
we are open and appreciated to accept prs and you could add yourself as an author to the package's README for sure.

before the pr is ready for review, feel free to ask me if you have any questions that i can help with.

**Comment by jthoward64** [comment]
> hey! that's awesome for doing such a great work even we didn't document much how to create a plugin yet.
> 
> we are open and appreciated to accept prs and you could add yourself as an author to the package's README for sure.
> 
> 
> 
> before the pr is ready for review, feel free to ask me if you have any questions that i can help with.

Thanks! Yeah I mostly just went off the navigation and tinybase plugins. One question I did have is how much should I worry about bundle size on the web UI? I'm probably gonna remove Ant Design anyways, but if the size is a big deal then it's definitely going away.

**Comment by jthoward64** [comment]
I don't have an emulator installed for Android atm, but I can't imagine there'd be any issues. Beyond that I'm pretty happy with the UI of the tool and everything seems pretty robust

vivaldi_wu8SEs8ebw

**Comment by jthoward64** [comment]
Found an issue, will fix sometime tonight or tomorrow (pardon the ping)

**Comment by Kudo** [comment]
> One question I did have is how much should I worry about bundle size on the web UI? I'm probably gonna remove Ant Design anyways, but if the size is a big deal then it's definitely going away.

bundle size should not be a concern.

- the first reason is that we have this convention for tree shaking, the dev tools code would be stripped from release build.
https://github.com/expo/dev-plugins/blob/7b8e083ae220941bd7ff84eee1fed2fcd4eb6db5/packages/react-navigation/src/index.ts#L1-L8
- secondly, we don't have any code to import/require code inside webui. that webui code will not be bundled inside your app.
- so webui code is only served through http dev server and loaded by browsers.

**Comment by jthoward64** [comment]
> > One question I did have is how much should I worry about bundle size on the web UI? I'm probably gonna remove Ant Design anyways, but if the size is a big deal then it's definitely going away.
> 
> bundle size should not be a concern.
> 
> * the first reason is that we have this convention for tree shaking, the dev tools code would be stripped from release build.
>   https://github.com/expo/dev-plugins/blob/7b8e083ae220941bd7ff84eee1fed2fcd4eb6db5/packages/react-navigation/src/index.ts#L1-L8
> * secondly, we don't have any code to import/require code inside webui. that webui code will not be bundled inside your app.
> * so webui code is only served through http dev server and loaded by browsers.

Gotcha, thanks

**Comment by jthoward64** [comment]
OK, pretty sure this is good for review/merge

File: packages/async-storage/webui/src/usePluginStore.ts [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+  const { message, notification } = App.useApp();
+
+  const [connected, setConnected] = useState(false);
+  useEffect(() => {
+    const interval = setInterval(() => {
+      setConnected(client?.isConnected() ?? false);
+    }, 1000);
+    return () => clearInterval(interval);
+  }, [client]);
```
Comment (Kudo) [inline packages/async-storage/webui/src/usePluginStore.ts:32]: is the loop only wait for the initial connected event?
just curious about it and think if that's anything we can help from the devtools plugin api level.

File: packages/async-storage/webui/package.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  "devDependencies": {
+    "@ant-design/icons": "^5.2.6",
+    "@babel/core": "^7.20.0",
+    "@react-native-async-storage/async-storage": "^1.21.0",
+    "@types/react": "~18.2.45",
+    "antd": "^5.10.2",
+    "expo": "50.0.0-preview.7",
+    "react": "18.2.0",
+    "react-dom": "18.2.0",
+    "@microlink/react-json-view": "^1.23.0",
```
Comment (Kudo) [inline packages/async-storage/webui/package.json:22]: is that `@microlink/react-json-view` still maintained so you using it here? do you recommend to migrate other react-json-view to `@microlink/react-json-view`?

Comment (jthoward64) [reply packages/async-storage/webui/package.json:22]: See the [message](https://github.com/mac-s-g/react-json-view?tab=readme-ov-file#%EF%B8%8F-this-project-is-no-longer-being-maintained-%EF%B8%8F) from the original project. Looks like there's also uiwjs's [react-json-view](https://github.com/uiwjs/react-json-view) which I didn't really look into. Not sure how well maintained it is, but it seems stable at the very least

File: packages/async-storage/webui/src/usePluginStore.ts (hunk 1 above)
Comment (jthoward64) [reply packages/async-storage/webui/src/usePluginStore.ts:32]: Pretty much yeah, the only expansion I can think of would be something like a `useConnection` hook that reports information about the connection (could be just isConnected, but maybe also error states?). Probably a cleaner way to handle the check with the current API though, I'll give it some thought

File: packages/async-storage/webui/package.json (hunk 2 above)
Comment (Kudo) [reply packages/async-storage/webui/package.json:22]: wow just check the [comparison in react-json-view](https://github.com/uiwjs/react-json-view?tab=readme-ov-file#size-and-dependencies) and find the [react-inspector](https://github.com/storybookjs/react-inspector). that's more promising to me, i will try to migrate react-inspector for other plugins.

File: packages/async-storage/webui/src/usePluginStore.ts (hunk 1 above)
Comment (Kudo) [reply packages/async-storage/webui/src/usePluginStore.ts:32]: i will also think for a proper solution for this, trying to re-render the hook when it's connected.

in the meantime, the downside of setInterval here is that even after connected, there is still a 1s loop running.
how about this?
```suggestion
  useEffect(() => {
    let interval = setInterval(() => {
      if (client?.isConnected()) {
        if (interval != null) {
          clearInterval(interval);
          interval = null;
        }
        setConnected(true);
      }
    }, 1000);
    return () => { if (interval != null) { clearInterval(interval); interval = null; } };
  }, [client]);
```

Comment (jthoward64) [reply packages/async-storage/webui/src/usePluginStore.ts:32]: I like it, yeah probably should've gone for that from the start tbh

File: packages/async-storage/webui/package.json (hunk 2 above)
Comment (jthoward64) [reply packages/async-storage/webui/package.json:22]: Agreed, I'll switch this PR over to it

File: packages/async-storage/webui/src/usePluginStore.ts (hunk 1 above)
Comment (Kudo) [reply packages/async-storage/webui/src/usePluginStore.ts]: i've revisited my api implementation. when `useDevToolsPluginClient()` returns, the websocket will be connected already. so we don't really need the timer to wait for connection.

i try to clone your branch and modify with the following patch. what do you think and is that aligned to your design?
```diff
diff --git a/packages/async-storage/webui/src/usePluginStore.ts b/packages/async-storage/webui/src/usePluginStore.ts
index ab9b502..319a2eb 100644
--- a/packages/async-storage/webui/src/usePluginStore.ts
+++ b/packages/async-storage/webui/src/usePluginStore.ts
@@ -23,24 +23,10 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const { message, notification } = App.useApp();
 
-  const [connected, setConnected] = useState(false);
-  useEffect(() => {
-    let interval = setInterval(() => {
-      if (client?.isConnected()) {
-        if (interval != null) {
-          clearInterval(interval);
-          interval = null;
-        }
-        setConnected(true);
-      }
-    }, 1000);
-    return () => { if (interval != null) { clearInterval(interval); interval = null; } };
-  }, [client]);
-
   const [entries, setEntries] = useState<readonly { key: string; value: string | null }[]>([]);
 
   const update = useCallback(async () => {
-    if (!client?.isConnected()) {
+    if (client?.isConnected() === false) {
       message.error('Not connected to host');
       return;
     }
@@ -53,12 +39,12 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const set = useCallback(
     async (key: string, value: string) => {
-      if (!client?.isConnected()) {
+      if (client?.isConnected() === false) {
         message.error('Not connected to host');
         return;
       }
       try {
-        return client.sendMessage('set', {
+        return client?.sendMessage('set', {
           key,
           value,
         });
@@ -71,7 +57,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
   const remove = useCallback(
     async (key: string) => {
-      if (!client?.isConnected()) {
+      if (client?.isConnected() === false) {
         message.error('Not connected to host');
         return;
       }
@@ -87,7 +73,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
   );
 
   useEffect(() => {
-    if (!client?.isConnected()) {
+    if (client?.isConnected() === false) {
       notification.error({
         message: 'Critical error:\nNot connected to host - please reload the page',
       });
@@ -98,7 +84,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(
+        client?.addMessageListener(
           methodAck.getAll,
           ({ result }: { result: readonly KeyValuePair[] }) => {
             setEntries(result.map(([key, value]) => ({ key, value })));
@@ -111,7 +97,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(methodAck.set, () => {
+        client?.addMessageListener(methodAck.set, () => {
           update();
         })
       );
@@ -121,7 +107,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
 
     try {
       subscriptions.push(
-        client.addMessageListener(methodAck.remove, () => {
+        client?.addMessageListener(methodAck.remove, () => {
           update();
         })
       );
@@ -130,7 +116,7 @@ export function usePluginStore(onError: (error: unknown) => void) {
     }
 
     subscriptions.push(
-      client.addMessageListener('error', ({ error }: { error: unknown }) => {
+      client?.addMessageListener('error', ({ error }: { error: unknown }) => {
         onError(error);
       })
     );
@@ -151,6 +137,6 @@ export function usePluginStore(onError: (error: unknown) => void) {
     update,
     set,
     remove,
-    ready: connected,
+    ready: client?.isConnected() === true,
   };
 }
```

Comment (jthoward64) [reply packages/async-storage/webui/src/usePluginStore.ts]: Yeah that looks good! Probably a better DX too. The only other weak point of the API I noticed (while I'm thinking of it) is in error handling/reporting, but realistically this is a dev-mode only package so it's probably fine. Thanks for the support on this!

File: packages/async-storage/webui/package.json (hunk 2 above)
Comment (jthoward64) [reply packages/async-storage/webui/package.json:22]: Looking more at it I'm not actually sure that react-inspector is a good fit as it doesn't seem to really support editing. I think as far as this plugin goes sticking with `@microlink/react-json-view` would be best, and maybe switch over to [react-json-view](https://uiwjs.github.io/react-json-view/) if there are any issues (and once it's 2.0 comes out of alpha). But the usage is pretty limited so I don't really see any reason it should break (outside of a React update I suppose)

**Review by Kudo — APPROVED** [review]
walking through the code again and everything looks good to me. i'm merging this pr and publish the package soon. thanks again for your awesome work for this plugin 👏

**Comment by Kudo** [comment]
`@dev-plugins/async-storage@0.0.1` published 🚀

**Comment by jthoward64** [comment]
> walking through the code again and everything looks good to me. i'm merging this pr and publish the package soon. thanks again for your awesome work for this plugin 👏

Of course; thanks for the advice on it!

**Pull request merged**